Resolve issues with Zygote.jl and `mean_vector`

soldasim opened the pull request:
<!-- Comment lines like this one will remain invisible -->

<!-- Thank you for your contribution! Please fill in this template so that we
can understand your intent and the proposed changes. If anything about this
template is unclear, just mention it! -->

**Summary**
<!-- Summary of what & why - explain your motivation and/or link to any GitHub issues this relates to -->

There is an issue with Zygote.jl not being able to correctly differentiate through `mean_vector`. It is described in [the docstring of `CustomMean`](https://github.com/JuliaGaussianProcesses/AbstractGPs.jl/blob/8d431a2ed31e3bf937640948c87aa85948ab8688/src/mean_function.jl#L42C1-L66C60). It proposes to the user to implement custom methods for their particular mean function if they wish to use Zygote.jl. This seems unnecessarily clumsy.

**Proposed changes**
<!-- Large PRs should ideally be preceded by a design discussion on a separate issue! -->

<!-- A clear and concise description of the contents of this pull request. -->

Correct me if I'm wrong, but I think this issue can be resolved by simply defining special methods for `ColVecs` and `RowVecs` as proposed in this PR. This way, Zygote.jl should work with `CustomMean` out of the box.

**What alternatives have you considered?**
<!-- A clear and concise description of any alternative solutions or features you've considered. -->

nothing

**Breaking changes**
<!-- If this PR breaks backwards-compatibility, please start the PR title with `**BREAKING**`! -->
<!-- In this section, describe any changes that are not backwards-compatible, -->
<!-- why it is worth breaking backwards compatiblity, -->
<!-- and how a user would have to address these changes in their downstream code. -->

This should not introduce any breaking changes to users who already resolved the issue by defining their own methods as described in the current docstring.

willtebbutt commented:
Thanks for this @soldasim . Could you add a couple of tests to check that it's working correctly?

I believe that the original reason that we introduced this was precisely because Zygote struggles with the kind of implementation that you propose (really bad performance), so we opted to not provide implementations in general. As such, it would also be good to see a benchmark that shows that Zygote's performance isn't completely terrible for this implementation.

soldasim commented:
Hi @willtebbutt,

I've added some a unit test in `test/mean_function.jl` for the two new methods.

I've also added [a problem](https://discourse.julialang.org/t/zygote-gradient-does-not-work-with-abstractgps-custommean/87815) that I've had in the past with Zygote as a test. This ensures that the new methods really fix the issue. I cannot be sure that different issues with Zygote.jl gradients could not exist, which may not be fixed by the two new methods though.

soldasim commented:
I have tested the performance on the following benchmark;

<details>
  <summary>benchmark</summary>
  
```julia
using AbstractGPs
using Random
using BenchmarkTools
using Zygote

rng = MersenneTwister(123456)
N, D = 5, 3
xD_colvecs = ColVecs(randn(rng, D, N))
xD_rowvecs = RowVecs(randn(rng, N, D))

X = randn(rng, D, N)
y = randn(rng, N)

foo_mean = x -> sum(abs2, x)

function construct_finite_gp(X, lengthscale, noise)
    mean = CustomMean(foo_mean)
    kernel = with_lengthscale(Matern52Kernel(), lengthscale)
    return GP(mean, kernel)(X, noise)
end

function loglike(lengthscale, noise)
    gp = construct_finite_gp(X, lengthscale, noise)
    return logpdf(gp, y)
end

# # User-defined methods.
# # Comment/Uncomment to compare the proposed implementation with user-defined methods.
# # (And restart REPL.)
# AbstractGPs.mean_vector(m::CustomMean{typeof(foo_mean)}, x::ColVecs) = vec(sum(abs2, x.X; dims=1))
# AbstractGPs.mean_vector(m::CustomMean{typeof(foo_mean)}, x::RowVecs) = vec(sum(abs2, x.X; dims=2))

display(@btime Zygote.gradient(n -> loglike(1., n), 1.))
display(@btime Zygote.gradient(l -> loglike(l, 1.), 1.))
```

</details>

Results with the proposed implementation:

```julia-repl
julia> 

  56.417 μs (697 allocations: 73.02 KiB)
(6.037955692260997,)
  56.709 μs (697 allocations: 73.02 KiB)
(5.834500256839596,)

julia> 
```

Results with the user-defined functions:

```julia-repl
julia> 

  55.625 μs (679 allocations: 69.52 KiB)
(6.037955692260998,)
  55.917 μs (679 allocations: 69.52 KiB)
(5.834500256839596,)

julia> 
```

At least in this case, the proposed implementation is as efficient as defining custom functions according to the old docstring.

willtebbutt reviewed (commented):
This looks to be nearly there now. Could you please bump the patch version, and also move the tests that you've added to `zygote.jl` in to `test/mean_function.jl` instead?

willtebbutt reviewed (approved):
Thanks @soldasim . Docs not building does not appear to have anything to do with this PR, so I'm going to merge + tag.
